application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml

```diff
@@ -61,5 +61,104 @@
+  &lt;select class="$class form-control"&gt;
+    #foreach($option in $options)
+      &lt;option value="$option"&gt;$option&lt;/option&gt;
+    #end
+  &lt;/select&gt;
+#end
+
+#macro(fieldsSelect $fields)
+  &lt;select class="form-control analyticsApplicationRowEvolutionField"&gt;
+    #foreach($key in $fields.keySet())
+      &lt;option value="$fields.get($key)"&gt;
+        $services.localization.render('analytics.rowEvolution.field.' + $key)
+      &lt;/option&gt;
+    #end
+  &lt;/select&gt;
+#end
+
+#macro(modalRowEvolution $fields $macroName)
+#set ($requestPageURL = $xwiki.getURL('Analytics.Code.Macros.RowEvolutionJson'))
+{{html clean=false}}
+&lt;div class="modal fade analyticsRowEvolutionModal" tabindex="-1" role="dialog" aria-labelledby="chartModalLabel"
+     aria-hidden="true"&gt;
+  &lt;div class="modal-dialog modal-lg"&gt;
+    &lt;div class="modal-content"&gt;
+      &lt;div class="modal-header"&gt;
+        &lt;button type="button" class="close" data-dismiss="modal" aria-label="Close"&gt;
+          &lt;span aria-hidden="true"&gt;&amp;times;&lt;/span&gt;
+        &lt;/button&gt;
+        &lt;h4 class="modal-title" class="chartModalLabel"&gt;
+          $services.localization.render('analytics.rowEvolution.label')
+        &lt;/h4&gt;
+      &lt;/div&gt;
+      &lt;div class="modal-body"&gt;
+        &lt;label for="analyticsApplicationModalPeriodSelect"&gt;
+          $services.localization.render('analytics.rowEvolution.period')
+        &lt;/label&gt;
+        &lt;select class="analyticsApplicationModalPeriodSelect form-control"&gt;
+          #durationOption("Day")
+          #durationOption("Week")
+          #durationOption("Month")
+          #durationOption("Year")
+        &lt;/select&gt;
+        &lt;div class="analyticsRowEvolutionOptions" data-macro-name=$macroName data-request-page=$requestPageURL data-period='day' data-date='30'&gt;
+          #durationSelectList("analyticsApplicationRowEvolutionDurationDay" [30, 8, 60, 90, 180, 365, 500])
+          #durationSelectList("analyticsApplicationRowEvolutionDurationWeek" [4, 12, 26, 52, 104, 500])
+          #durationSelectList("analyticsApplicationRowEvolutionDurationMonth" [3, 6, 12, 24, 36, 120])
+          #durationSelectList("analyticsApplicationRowEvolutionDurationYear" [3, 5, 10])
+        &lt;/div&gt;
+        &lt;label for="analyticsApplicationRowEvolutionField"&gt;
+          $services.localization.render('analytics.rowEvolution.field')
+        &lt;/label&gt;
+        #fieldsSelect($fields)
+        &lt;canvas class="analyticsChartisplay"&gt;&lt;/canvas&gt;
+      &lt;/div&gt;
+      &lt;div class="modal-footer"&gt;
+        &lt;button type="button" class="btn btn-default" data-dismiss="modal"&gt;
+          $services.localization.render('analytics.rowEvolution.close')
+        &lt;/button&gt;
+      &lt;/div&gt;
+    &lt;/div&gt;
+  &lt;/div&gt;
+&lt;/div&gt;
+{{/html}}
+#end
+
+#set ($analyticsId = -1)
+#macro (getAnalyticsMacroId $macro)
```

Comment:
`#macro (getAnalyticsMacroId $macroName $macroId)`
Because right now you don't know when this macro is called from where the `$macroId` comes. I know it's a little strange, since it gives the feeling that you add the return in the parameters list, but it's more like a helper to know where some variables are created